### PR TITLE
#1564 Improve configuration around readAfterReplicaAcknowledged()

### PR DESCRIPTION
**Pull request by RobAustin**
Relates to  and https://github.com/ChronicleEnterprise/Chronicle-Queue-Enterprise/pull/704

### CODE REVIEW

File: src/main/java/net/openhft/chronicle/queue/impl/WireStorePool.java [hunk 1]
```diff
@@ -32,7 +32,7 @@ public class WireStorePool extends SimpleCloseable {
     private final WireStoreSupplier supplier;
     private final StoreFileListener storeFileListener;
 
-    private WireStorePool(@NotNull WireStoreSupplier supplier, StoreFileListener storeFileListener) {
+    WireStorePool(@NotNull WireStoreSupplier supplier, StoreFileListener storeFileListener) {
```
Comment (tgd) [inline src/main/java/net/openhft/chronicle/queue/impl/WireStorePool.java]: Is this change unrelated?

File: src/main/java/net/openhft/chronicle/queue/ExcerptTailer.java [hunk 2]
```diff
@@ -290,4 +319,13 @@ default long excerptsInCycle(int cycle) {
     default @NotNull ExcerptTailer afterWrittenMessageAtIndex(@NotNull ChronicleQueue queue, long index) {
         throw new UnsupportedOperationException("todo");
     }
+
+    interface AcknowledgedIndexReplicatedCheck {
+        /**
+         * @param index           the index of the next message for the tailer to read
```
Comment (tgd) [inline src/main/java/net/openhft/chronicle/queue/ExcerptTailer.java:329]: I would recommend adding a Javadoc here letting implementers know they need to be careful around the roll when calculating delta between index values and should lean on cycle calculations as needed if correctness in terms of index delta calculation is critical for them during the roll

File: src/main/java/net/openhft/chronicle/queue/impl/single/StoreTailer.java [hunk 3]
```diff
@@ -1194,10 +1210,16 @@ void releaseStore() {
     @Override
     public void readAfterReplicaAcknowledged(final boolean readAfterReplicaAcknowledged) {
         throwExceptionIfClosed();
-
         this.readAfterReplicaAcknowledged = readAfterReplicaAcknowledged;
     }
 
+
+    @Override
+    public void acknowledgedIndexReplicatedCheck(final AcknowledgedIndexReplicatedCheck acknowledgedIndexReplicatedCheck) {
```
Comment (tgd) [inline src/main/java/net/openhft/chronicle/queue/impl/single/StoreTailer.java]: Would be good to enforce non-null on the check

File: src/main/java/net/openhft/chronicle/queue/impl/single/StoreTailer.java [hunk 4]
```diff
@@ -469,10 +469,25 @@ private boolean inACycle2(boolean includeMetaData, Wire wire, Bytes<?> bytes) th
...
         final long lastSequenceAck = queue.lastAcknowledgedIndexReplicated();
         final long index = index();
-        return index > lastSequenceAck;
+        if (lastSequenceAck == -1)
+            return true; // don't allow the message to be read
+        if (index == Long.MIN_VALUE)
+            return false; // allow the message to be read
+
+        // check both are in the same roll cycle
+        if (queue.rollCycle().toCycle(index) != queue.rollCycle().toCycle(lastSequenceAck))
```
Comment (tgd) [inline src/main/java/net/openhft/chronicle/queue/impl/single/StoreTailer.java:484]: If the implementation doesn't support calculating the delta across roll cycles I reckon we should call that out on the javadoc (if not already done)

**Comment by tgd** [comment]
Hi @RobAustin thanks! Approved this with comments - note there are some build failures - not sure why.

**Review by tgd — APPROVED** [review]

File: src/main/java/net/openhft/chronicle/queue/impl/single/StoreTailer.java (hunk 4 above)
Comment (RobAustin) [reply src/main/java/net/openhft/chronicle/queue/impl/single/StoreTailer.java:484]: Agreed

File: src/main/java/net/openhft/chronicle/queue/ExcerptTailer.java [hunk 5]
```diff
@@ -290,4 +319,13 @@ default long excerptsInCycle(int cycle) {
...
+         * @param lastSequenceAck the last index that has been acknowledged as replicated
+         * @return true if you are ok for the tailer to read the message at {@code index}
+         */
+        boolean acknowledgedIndexReplicatedCheck(long index, long lastSequenceAck);
```
Comment (JerryShea) [inline src/main/java/net/openhft/chronicle/queue/ExcerptTailer.java:329]: please add `byte remoteHostId`

File: src/test/java/net/openhft/chronicle/queue/impl/single/AcknowledgedIndexReplicatedCheckTest.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+import org.junit.Test;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+public class AcknowledgedIndexReplicatedCheckTest {
+
+    @Test
+    public void testReadBeforeAcknowledgment() throws IOException {
```
Comment (JerryShea) [inline src/test/java/net/openhft/chronicle/queue/impl/single/AcknowledgedIndexReplicatedCheckTest.java]: please add to LastAcknowledgedTest

**Review by peter-lawrey — APPROVED** [review]

File: src/main/java/net/openhft/chronicle/queue/ExcerptTailer.java [hunk 7]
```diff
@@ -211,6 +211,41 @@ default ExcerptTailer afterLastWritten(ChronicleQueue queue) {
...
+     * However, if you wish to tolerate a number of in-flight or pending messages before the message.,
+     * You can set this to a different implementation. For example, if you wish to tolerate one unacknowledged message, set {@code tolerateNumberOfUnAckedMessages} to 1, as shown below:
+     * <pre>
+     * boolean acknowledgedIndexReplicatedCheck(long index, long lastSequenceAck) {
+     *     int tolerateNumberOfUnAckedMessages = 1;
+     *     return index <= lastSequenceAck + tolerateNumberOfUnAckedMessages;
+     * }
+     * </pre>
+     * <p>
+     * This feature is exclusive to Enterprise Queue and is disabled when accessing a queue file role.
```
Comment (JerryShea) [inline src/main/java/net/openhft/chronicle/queue/ExcerptTailer.java:239]: "when accessing a queue file role" -> "disabled across queue file *roll*"?

**Review by JerryShea — APPROVED** [review]
approved - one minor javadoc comment

**Pull request merged**